drivers/counter/rtc_mcp7940n.c

```diff
@@ -0,0 +1,771 @@
+
+	int rc = i2c_write_read(cfg->i2c_dev, cfg->addr,
+				&addr, sizeof(addr),
+				&data->registers, RTC_TIME_REGISTERS_SIZE);
+
+	if (rc >= 0) {
+		*unix_time = decode_rtc(dev);
+	}
+
+	return rc;
+}
+
+/** @brief Write a single register to MCP7940N
+ *
+ * @param dev the MCP7940N device pointer.
+ * @param addr register address.
+ * @param value Value that will be written to the register.
+ *
+ * @retval return 0 on success, or a negative error code from an I2C
+ * transaction or invalid parameter.
+ */
+static int write_register(const struct device *dev, enum mcp7940n_register addr, uint8_t value)
+{
+	const struct mcp7940n_config *cfg = dev->config;
+	int rc = 0;
+
+	uint8_t time_data[2] = {addr, value};
+
+	rc = i2c_write(cfg->i2c_dev, time_data, sizeof(time_data), cfg->addr);
+
+	return rc;
+}
+
+/** @brief Write a full time struct to MCP7940N registers.
+ *
+ * @param dev the MCP7940N device pointer.
+ * @param addr first register address to write to, should be REG_RTC_SEC,
+ * REG_ALM0_SEC or REG_ALM0_SEC.
+ * @param size size of data struct that will be written.
+ *
+ * @retval return 0 on success, or a negative error code from an I2C
+ * transaction or invalid parameter.
+ */
+static int write_data_block(const struct device *dev, enum mcp7940n_register addr, uint8_t size)
+{
+	struct mcp7940n_data *data = dev->data;
+	const struct mcp7940n_config *cfg = dev->config;
+	int rc = 0;
+	uint8_t time_data[MAX_WRITE_SIZE + 1];
+	uint8_t *write_block_start;
+
+	if (size > MAX_WRITE_SIZE) {
+		return -EINVAL;
+	}
+
+	if (addr >= REG_INVAL) {
+		return -EINVAL;
+	}
+
+	if (addr == REG_RTC_SEC) {
+		write_block_start = (uint8_t *)&data->registers;
+	} else if (addr == REG_ALM0_SEC) {
+		write_block_start = (uint8_t *)&data->alm0_registers;
+	} else if (addr == REG_ALM1_SEC) {
+		write_block_start = (uint8_t *)&data->alm1_registers;
+	} else {
+		return -EINVAL;
+	}
+
+	/* Load register address into first byte then fill in data values */
+	time_data[0] = addr;
+	memcpy(&time_data[1], write_block_start, size);
+
+	rc = i2c_write(cfg->i2c_dev, time_data, size + 1, cfg->addr);
+
+	return rc;
+}
+
+/** @brief Sets the correct weekday.
+ *
+ * If the time is never set then the device defaults to 1st January 1970
+ * but with the wrong weekday set. This function ensures the weekday is
+ * correct in this case.
+ *
+ * @param dev the MCP7940N device pointer.
+ * @param unix_time pointer to unix time that will be used to work out the weekday
+ *
+ * @retval return 0 on success, or a negative error code from an I2C
+ * transaction or invalid parameter.
+ */
+static int set_day_of_week(const struct device *dev, time_t *unix_time)
+{
+	struct mcp7940n_data *data = dev->data;
+	struct tm time_buffer = { 0 };
+	int rc = 0;
+
+	gmtime_r(unix_time, &time_buffer);
+
+	if (time_buffer.tm_wday != 0) {
+		data->registers.rtc_weekday.weekday = time_buffer.tm_wday;
+		rc = write_register(dev, REG_RTC_WDAY,
+			*((uint8_t *)(&data->registers.rtc_weekday)));
```

Comment:
Question: Is there an established way in the zephyr code-base to get rid of this type of casting?

I would also be ok with keeping it as it is now.

Reply:
@carlescufi @mbolivar could you comment on the above?

The underlying device registers are contained within 8-bit bitfields, with the different parts of the register occupying 1 or more bits within. So either a pointer is passed to the write register function, or the bitfield value is extracted using a cast. Is there a preference for how this should be handled?

Reply:
> @carlescufi @mbolivar could you comment on the above?

I'm not aware of a convention and this seems OK to me too.

Reply:
OK, thx.